src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -395,12 +395,19 @@ public static AbstractMemorySegmentImpl ofBuffer(ByteBuffer bb) {
 
         MemoryScope bufferScope = new MemoryScope(bb, null);
         int size = limit - pos;
+
+        MemorySegment bufferSegment = (MemorySegment)nioAccess.bufferSegment(bb);
+        int modes = bufferSegment == null ?
+                defaultAccessModes(size) : bufferSegment.accessModes();
+        if (bb.isReadOnly()) {
+            modes &= ~WRITE;
+        }
         if (base != null) {
-            return new HeapMemorySegmentImpl<>(bbAddress + pos, () -> (byte[])base, size, defaultAccessModes(size), Thread.currentThread(), bufferScope);
+            return new HeapMemorySegmentImpl<>(bbAddress + pos, () -> (byte[])base, size, modes, Thread.currentThread(), bufferScope);
         } else if (unmapper == null) {
-            return new NativeMemorySegmentImpl(bbAddress + pos, size, defaultAccessModes(size), Thread.currentThread(), bufferScope);
+            return new NativeMemorySegmentImpl(bbAddress + pos, size, modes, Thread.currentThread(), bufferScope);
```

Comment:
I think this and prior code can result in an integrity issue with segment -> buffer -> segment. One can close the original segment while retaining access on another thread to the segment obtained from the buffer.

Consider the following example:
```
        CountDownLatch a = new CountDownLatch(1);
        CountDownLatch b = new CountDownLatch(1);
        CompletableFuture<?> r;
        try (MemorySegment s1 = MemorySegment.allocateNative(intArrayLayout)) {
            r = CompletableFuture.runAsync(() -> {
                try {
                    ByteBuffer bb = s1.asByteBuffer();

                    MemorySegment s2 = MemorySegment.ofByteBuffer(bb);
                    a.countDown();

                    try {
                        b.await();
                    } catch (InterruptedException e) {
                    }

                    MemoryAddress base = s2.baseAddress();
                    intElemHandle.set(base, 1L, -42);
                } catch (RuntimeException e) {
                    e.printStackTrace();
                    throw e;
                }
            });

            a.await();
            MemoryAddress base = s1.baseAddress();
            intElemHandle.set(base, 1L, 42);
        }

        b.countDown();
        r.get();
```

Reply:
Yeah. Instead of always using Thread.currentThread(), I think we can use the owner thread of `bufferSegment` if it's not null. That makes the above example fail.

Reply:
Good point, I will fix this. It's interesting how all these issues were latent in the code and having a more direct implementation for the various segments made them pop out more :-)